mantidproject/mantid
# Fix loading old muon NeXus files without frames_good

Fix a bug seen in MuonAnalysis where attempting to load certain old muon NeXus files failed and caused the interface to lock up.

Cause: the file did not contain number of good frames, so no value was written to `goodfrm` in the workspace's run object. This made subsequent algorithms (`ApplyDeadTimeCorr`) fail.
- If `run/instrument/beam/frames_good` is not present, use `run/instrument/beam/frames` instead and assume all were good. Warn the user in the log that this is happening. 
- Added a test and updated documentation
- In `MuonProcess`, ensured algorithm fails fast if dead time correction fails. This avoids the interface locking up as was seen, and gives the user a clearer error message.

**To test:**
- The data files from the issue can be found at `\\olympic\babylon5\Scratch\TomPerkins\Data\keren_data`. These are old files that have been converted to NeXus and have no `run/instrument/beam/frames_good` field.
- An alternative data file to try is `EMU30604.nxs` from the system test data.
1. _From script:_
   - Load one of these files: 
   
   ``` python
   LoadMuonNexus(Filename='path/to/53590.NXS', OutputWorkspace='ws', DeadTimeTable='dt')
   ```
   - Look in the sample logs of the loaded workspace `ws` and check `goodfrm` is present. 
   - Check the log for a warning that `run/instrument/beam/frames` was used instead of   `run/instrument/beam/frames_good`.
   - Attempt to apply dead time correction (should not give an error):
   
   ``` python
   ApplyDeadTimeCorr(InputWorkspace='ws', DeadTimeTable='dt', OutputWorkspace='corr53590')
   ```
2. _Using GUI:_
   - Set facility to ISIS, open _Interfaces/Muon/Muon Analysis_ and set the "Dead Time Correction" dropdown to "From data file"
   - Select one of the files with "Browse" and load it
   - No error should be seen (should see a plot appear)

Fixes #16723.

[Release notes](https://github.com/mantidproject/mantid/blob/a147f4699abc655164f7cbf94bfb5976bdaf9832/docs/source/release/v3.8.0/muon.rst#algorithms) have been updated.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
